### PR TITLE
os, build/configs: add config for common data sections size, clean up xip elf code

### PR DESCRIPTION
**Pull request by abhishek-samsung**
1. added config for common ram size when using xip loadable apps
2. calculated heap allocation for kernel based on app ram size configs
3. added checks for 4KB (l2 page alignment) aligned for app flash and ram partitions

### CODE REVIEW

File: build/configs/rtl8730e/loadable_apps_xip/defconfig [hunk 1]
```diff
@@ -8,6 +8,7 @@
 #
 # CONFIG_EXPERIMENTAL is not set
 # CONFIG_DEFAULT_SMALL is not set
+CONFIG_DOCKER_VERSION="2.0.0"
```
Comment (sunghan-chang) [inline build/configs/rtl8730e/loadable_apps_xip/defconfig:11]: Do we still need 2.0.0 with loadable XIP?

File: build/configs/rtl8730e/loadable_apps_xip/defconfig [hunk 2]
```diff
@@ -27,6 +28,7 @@ CONFIG_NUM_APPS=2
 CONFIG_SUPPORT_COMMON_BINARY=y
 CONFIG_COMMON_BINARY_VERSION="200204"
 CONFIG_COMMON_BINARY_NAME="common"
+CONFIG_COMMON_BIN_DYN_RAMSIZE=524288
```
Comment (sunghan-chang) [inline build/configs/rtl8730e/loadable_apps_xip/defconfig:31]: What does it mean? A common binary uses separate heap?

File: build/configs/rtl8730e/loadable_apps_xip/defconfig (hunk 1 above)
Comment (abhishek-samsung) [reply build/configs/rtl8730e/loadable_apps_xip/defconfig:11]: It doesn't require newer toolchain. We can use older toolchain also.

File: build/configs/rtl8730e/loadable_apps_xip/defconfig (hunk 2 above)
Comment (abhishek-samsung) [reply build/configs/rtl8730e/loadable_apps_xip/defconfig:31]: Its the space in RAM we allocate for data and bss sections of common binary.

Comment (abhishek-samsung) [reply build/configs/rtl8730e/loadable_apps_xip/defconfig:31]: I am currently using the same configs CONFIG_*_BIN_DYN_RAMSIZE for the space allocated for each app or common in the ram. If its common that space is used only for data and bss. If its app, its used for data, bss and heap. I think using the name DYN_RAMSIZE for common is wrong for common. I will change it to CONFIG_COMMON_BIN_STATIC_RAMSIZE

Comment (sunghan-chang) [reply build/configs/rtl8730e/loadable_apps_xip/defconfig]: Yes, we need to rename it as you said.
But do we allocate a data and bss in `CONFIG_APP1_BIN_DYN_RAMSIZE`??

Comment (abhishek-samsung) [reply build/configs/rtl8730e/loadable_apps_xip/defconfig]: Yes, data and bss for apps are allocated in CONFIG_APP1_BIN_DYN_RAMSIZE itself. We first allocate data and bss, then rest of it is allocated as heap.

File: build/configs/rtl8730e/loadable_apps_xip/defconfig (hunk 1 above)
Comment (edwakuwaku) [reply build/configs/rtl8730e/loadable_apps_xip/defconfig:11]: I recommend to use new toolchain, although there is no significant improvement from overall code size.
But along the way the toolchain version improve (ie. v6.3.1 -> v10.3.1), it includes some bug fix and optimization.

File: os/tools/mkldscript.py [hunk 3]
```diff
@@ -79,26 +79,42 @@
 
 ram_end = int(CONFIG_RAM_SIZE) + int(CONFIG_RAM_START, 16)
 
+if ram_end % 4096 != 0 :
+    print("RAM end should be 4KB aligned")
```
Comment (r-prabu) [inline os/tools/mkldscript.py:83]: Is the 4KB check is coupled with l2 page size?

**Review by r-prabu — APPROVED** [review]

File: os/tools/mkldscript.py (hunk 3 above)
Comment (abhishek-samsung) [reply os/tools/mkldscript.py:83]: Yes. For apps we always allocate l2 pages.

File: build/configs/rtl8730e/loadable_apps_xip/defconfig (hunk 2 above)
Comment (abhishek-samsung) [reply build/configs/rtl8730e/loadable_apps_xip/defconfig]: As discussed, we will take up changing the desing to remove fixed CONFIG_COMMON_BIN_STATIC_RAMSIZE

**Review by ewoodev — APPROVED** [review]

**Review by sunghan-chang — APPROVED** [review]

**Review by r-prabu — APPROVED** [review]

**Review by ritesh55555 — APPROVED** [review]

**Pull request merged**